user status: Focus status input box on modal display.

agrawal-d opened the pull request:
The code for focusing the text box already exists - but it's buggy. Adding **at least** 30ms of timeout before calling 'focus()' fixes this. Checked for [0...20] ms, doesn't work - 30ms is the minimum timeout required for this to work.

<!-- What's this PR for?  (Just a link to an issue is fine.) -->


**Testing Plan:** <!-- How have you tested? -->
Manually, in the browser UI.



**GIFs or Screenshots:** <!-- If a UI change.  See:
  https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
  -->
![focus](https://user-images.githubusercontent.com/7714968/76140706-3b4b6300-6083-11ea-81a1-54790bd56adb.gif)


<!-- Also be sure to make clear, coherent commits:
  https://zulip.readthedocs.io/en/latest/contributing/version-control.html
  -->

timabbott commented:
Can you try doing a timeout with 0?  Generally, constants like 20ms here are almost certain to be buggy in some race condition where some thing takes more than 20ms, and so are a code smell.  

If all we need is to run this after the browser has finished rendering the modal, a timeout of 0 will work (and be reliable); if that doesn't work, we should investigate further how to set the focus in a reliable way.

agrawal-d commented:
The code for focusing the text box already exists - but it was
broken in commit 6696fda which adds a transition of 0.2s to 'visibility' - but 'visibility:hidden' elements can't be focused by javascript, so adding a 200ms timeout
to 'focus()' reliably fixes this. There is no other way to do this than a timeout. Any timeout lesser than 200ms is unreliable - it works sometimes, and sometimes it doesn't.

timabbott commented:
OK, that makes sense, but makes me wonder if the transition is the right approach.  What if we added an `on-visible` handler that sets the focus when the widget finishes becoming visible?  Then we wouldn't need to have these two 200ms values in different parts of the codebase that need to stay synchronized (which I'd want to clarify with comments if we were to do it that way).

@andersk FYI since this appears to be a side effect of one of your changes.

agrawal-d commented:
> What if we added an on-visible handler

@timabbott I searched, but it appears there is no handler for visibility, neither in pure JS nor in JQuery. 
I've added an explanatory comment on top of the change. 

Please re-review.

agrawal-d commented:
Provision failed on latest master :(

arpit551 commented:
@agrawal-d  Your branch is 3 commits behind zulip:master.

agrawal-d commented:
@arpit551 those 3 commits were merges some 15 minutes ago. Thanks for taking a look, I rebased again.

timabbott commented:
The test failure is at test coverage gap; I guess that may be my fault in changes we recently merged to master.  I'll take a look.
```
Mar 26 20:48:44 Starting node tests...
Mar 26 20:48:44 ERROR: static/js/server_events_dispatch.js no longer has complete node test coverage
Mar 26 20:48:44   Lines missing coverage: 191
```

timabbott commented:
@agrawal-d I think 07bca27fbc79a2b914649d88e459800f46a9ed46 was intended to fix this ; can you test whether that solution covers this as well?  I like the approach a lot better, certainly.

agrawal-d commented:
It does - and certainly, it's a much better fix!
